File: java/src/deepspace/GameUniverse.java
/*
 * Here comes the text of your license
 * Each line should be prefixed with  * 
 */
package deepspace;
import java.util.ArrayList;

public class GameUniverse {
    
    private static final int WIN = 10;
    
    private int currentStationIndex;
    private SpaceStation currentStation;
    private EnemyStarShip currentEnemy; 
    private ArrayList<SpaceStation> spaceStations;
    
    private int turns;
    private Dice dice = new Dice();
    private GameStateController gameState = new GameStateController();
    private CardDealer dealer;
    
    public GameUniverse(){
        turns = 0;
    }
    
    public CombatResult combat(){
        GameState state = gameState.getState();   
        if (state == GameState.BEFORECOMBAT || state == GameState.INIT)
            return combat(currentStation, currentEnemy);
        else
            return CombatResult.NOCOMBAT;
    }
    
    CombatResult combat(SpaceStation station, EnemyStarShip enemy){
        
        GameState state = gameState.getState();
        if (state == GameState.BEFORECOMBAT || state == GameState.INIT){
            GameCharacter ch = dice.firstShot();
            
            boolean enemyWins;
            CombatResult combatResult = CombatResult.STATIONESCAPES;
            
            if (ch == GameCharacter.ENEMYSTARSHIP){
                
                float fire = station.fire();
                ShotResult result = station.receiveShot(fire);
                
                if (result == ShotResult.RESIST){
                    fire = station.fire();
                    result = enemy.receiveShot(fire);
                    enemyWins = result == ShotResult.RESIST;
                }
                else{
                    enemyWins = true;
                }
            }else{
                float fire = station.fire();
                ShotResult result = enemy.receiveShot(fire);
                enemyWins = result == ShotResult.RESIST;
            }
            
            if (enemyWins){
                float s = station.getSpeed();
                boolean moves = dice.spaceStationMoves(s);
                
                if (moves){
                    station.move();
                    combatResult = CombatResult.STATIONESCAPES;
                }else{
                    Damage damage = enemy.getDamage();
                    station.setPendingDamage(damage);
                    combatResult = CombatResult.STATIONWINS;
                }
            }
            
            gameState.next(turns, spaceStations.size());
            return combatResult;
        }    
        else
            return CombatResult.NOCOMBAT;
            
    }
    
    public void discardHangar(){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).discardHangar();
        }
    }
    
    public void discardWeapon(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).discardWeapon(i);
        }
    }
    
    public void discardWeaponInHangar(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).discardWeaponInHangar(i);
        }
    }
    
    public void discardShieldBooster(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).discardShieldBooster(i);
        }   
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).discardShieldBoosterInHangar(i);
        }
    }
    
    public GameState getState(){
        return gameState.getState();
    }
    
    public GameUniverseToUI getUIversion(){
        return new GameUniverseToUI(currentStation, currentEnemy);
    }
    
    public boolean haveAWinner(){
        return spaceStations.get(currentStationIndex).getNMedals() == WIN;
    }
    
    public void init(ArrayList<String> names){
        if (getState() == GameState.CANNOTPLAY){
            spaceStations = new ArrayList<>();
            dealer = CardDealer.getInstance();
            
            for(String name: names){
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(name,supplies);
                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();
                
                Loot l = new Loot(0,nw,ns,nh,0);
                
                station.setLoot(l);
                spaceStations.add(station);
            }
            
            currentStationIndex = dice.whoStarts(names.size());
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();
            gameState.next(turns, spaceStations.size());
        }
    }
    
    public void mountWeapon(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).mountWeapon(i);
        }
    }
    
    public void mountShieldBooster(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            spaceStations.get(currentStationIndex).mountShieldBooster(i);
        }
    }
    
    public boolean nextTurn(){
        
        if (getState() == GameState.AFTERCOMBAT){
            boolean stationState = currentStation.validState();
            
            if (stationState){
                currentStationIndex = (currentStationIndex + 1) % spaceStations.size();
                turns += 1;
                
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                gameState.next(turns, spaceStations.size());
                return true;
            }
        }
        
        return false;
    }
}

File: ruby/lib/GameUniverse.rb
require_relative 'Dice'
require_relative 'SpaceStation'
require_relative 'GameStateController'
require_relative 'GameUniverseToUI'
require_relative 'GameState'
require_relative 'GameCharacter'
require_relative 'ShotResult'
require_relative 'CombatResult'
require_relative 'Loot'
require_relative 'SpaceCity'
require_relative 'BetaPowerEfficientSpaceStation'
require_relative 'PowerEfficientSpaceStation'

module Deepspace
  class GameUniverse
    @@WIN = 10
    
    def initialize()
      @currentStationIndex = nil
      @gameState = GameStateController.new
      @currentEnemy = nil
      @turns = 0
      @dice = Dice.new()
      @haveSpaceCity = false
    end
    
    def discardHangar
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].discardHangar
      end
    end
    
    def discardWeapon(i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].discardWeapon(i)
      end
    end
    
    def discardWeaponInHangar(i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].discardWeaponInHangar(i)
      end
    end
    
    def discardShieldBooster(i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].discardShieldBooster(i)
      end
    end
    
    def discardShieldBoosterInHangar(i)
      if state == GameState::INIT || state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].discardShieldBoosterInHangar(i)
      end
    end
    
    def mountWeapon(i)
      if state == GameState::INIT or state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].mountWeapon(i)
      end
    end
    
    def mountShieldBooster(i)
      if state == GameState::INIT || state == GameState::AFTERCOMBAT
        @spaceStations[@currentStationIndex].mountShieldBooster(i)
      end
    end
    
    
    def haveAWinner
      @spaceStations[@currentStationIndex].nMedals == @@WIN
    end
    
    def state
      @gameState.state
    end
    
    def getUIversion
      GameUniverseToUI.new(@currentStation, @currentEnemy)
    end
    
    def init(names)
      
      if state == GameState::CANNOTPLAY
        @spaceStations = Array.new
        dealer = CardDealer.instance 
        
        names.each{|name|
          supplies = dealer.nextSuppliesPackage
          station = SpaceStation.new(name, supplies)
          nh = @dice.initWithNHangars
          nw = @dice.initWithNWeapons
          ns = @dice.initWithNShields
          
          l = Loot.new(0, nw, ns, nh, 0)
          
          station.setLoot(l)
          @spaceStations.push(station)
          
        }
        
        @currentStationIndex = @dice.whoStarts(names.length)
        @currentStation = @spaceStations[@currentStationIndex]
        @currentEnemy = dealer.nextEnemy
        @gameState.next(@turns, @spaceStations.length)
      end
      
    end
    
    def nextTurn
      gameState = @gameState.state
      if gameState == GameState::AFTERCOMBAT
        stationState = @currentStation.validState
        if stationState
          @currentStationIndex = (@currentStationIndex +1) % @spaceStations.length
          @turns += 1
          
          @currentStation = @spaceStations[@currentStationIndex]
          @currentStation.cleanUpMountedItems
          dealer = CardDealer.instance 
          @currentEnemy = dealer.nextEnemy
          @gameState.next(@turns, @spaceStations.length)
          return true
        end
      end
      false
    end
    
    def combat
      state = @gameState.state
      
      if (state == GameState::BEFORECOMBAT) || (state == GameState::INIT)
        combatGo(@currentStation, @currentEnemy)
      else
        CombatResult::NOCOMBAT
      end
      
    end
    
    def combatGo(station, enemy)
      
      ch = @dice.firstShot
      
      #Establecemos el combate
      
      if ch == GameCharacter::ENEMYSTARSHIP
        fire = enemy.fire
        result = station.receiveShot(fire)
        
        if result == ShotResult::RESIST
          fire = station.fire
          result = enemy.receiveShot(fire)
          enemyWins = result == ShotResult::RESIST
        else  
          enemyWins = true
        end
      else
        fire = station.fire
        result = enemy.receiveShot(fire)
        enemyWins = result==ShotResult::RESIST
      end
      
      #Analizamos resultado del combate
      
      if enemyWins
        s = station.speed
        moves = @dice.spaceStationMoves(s)
        
        if moves 
          station.move
          combatResult=CombatResult::STATIONESCAPES
        else
          damage = enemy.damage
          station.setPendingDamage(damage)
          combatResult=CombatResult::ENEMYWINS
        end
      else
        aLoot = enemy.loot
        station.setLoot(aLoot)
        combatResult=CombatResult::STATIONWINSANDCONVERTS
        
        if aLoot.efficient
          makeStationEfficient
        elsif aLoot.spaceCity  
          createSpaceCity
        else  
          combatResult=CombatResult::STATIONWINSANDCONVERTS
        end
        
        
      end
      @gameState.next(@turns, @spaceStations.length)
      return combatResult
      
    end
    
    def createSpaceCity
      if not @haveSpaceCity
        city = SpaceCity(@spaceStation[@currentStationIndex], Array.new(@spaceStation).delete_at(@currentStationIndex))
        @spaceStation[@currentStationIndex] = city
        @haveSpaceCity = true
        puts "TRANSFORMATION TO SPACECITY FINISHED"
      end
    end
    
    def makeStationEfficient
      if @dice.extraEfficiency
        station = BetaPowerEfficientSpaceStation.new(@spaceStations[@currentStationIndex])
        puts "TRANSPORMATED TO BETAEFFICIENT"
      else
        station = PowerEfficientSpaceStation.new(@spaceStations[@currentStationIndex])
        puts "TRANSPORMATED TO EFFICIENT"
      end
      
      @spaceStation[@currentStationIndex] = station
      
    end
    
  end
end
